File: launcher/ClatterDrive.Launcher.Tests/LauncherViewModelTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.IO;

namespace ClatterDrive.Launcher.Tests;

[TestClass]
public sealed class LauncherViewModelTests
{
    [TestMethod]
    public void StartCommandReflectsValidation()
    {
        using var backend = new FakeBackendController();
        using var viewModel = new LauncherViewModel(backend);

        Assert.IsTrue(viewModel.StartCommand.CanExecute(null));

        viewModel.Port = 0;

        Assert.IsFalse(viewModel.StartCommand.CanExecute(null));
        viewModel.Start();
        Assert.AreEqual("Choose a port from 1 to 65535.", viewModel.Status);
        Assert.AreEqual(0, backend.StartCalls);
    }

    [TestMethod]
    public void StartStopUpdatesStateAndPassesSettings()
    {
        var tempDir = Path.Combine(Path.GetTempPath(), $"clatterdrive-launcher-test-{Guid.NewGuid():N}");
        try
        {
            using var backend = new FakeBackendController();
            using var viewModel = new LauncherViewModel(backend)
            {
                BackingDirectory = tempDir,
                Port = 8123,
                AudioMode = "off",
                CapacityGb = 24.0,
                FilesystemProfile = "ext4_like",
                StatePath = Path.Combine(tempDir, "state.json"),
            };

            viewModel.Start();
            backend.RaiseReady();

            Assert.IsTrue(Directory.Exists(tempDir));
            Assert.AreEqual(1, backend.StartCalls);
            Assert.IsTrue(viewModel.IsRunning);
            Assert.AreEqual("Running", viewModel.Status);
            Assert.AreEqual(8123, backend.LastSettings?.Port);
            Assert.AreEqual("off", backend.LastSettings?.AudioMode);
            Assert.AreEqual(24.0, backend.LastSettings?.CapacityGb);
            Assert.AreEqual("ext4_like", backend.LastSettings?.FilesystemProfile);
            Assert.AreEqual(Path.Combine(tempDir, "state.json"), backend.LastSettings?.StatePath);
            Assert.IsFalse(viewModel.StartCommand.CanExecute(null));
            Assert.IsTrue(viewModel.StopCommand.CanExecute(null));

            viewModel.Stop();

            Assert.AreEqual(1, backend.StopCalls);
            Assert.IsFalse(viewModel.IsRunning);
            Assert.AreEqual("Stopped", viewModel.Status);
        }
        finally
        {
            if (Directory.Exists(tempDir))
            {
                Directory.Delete(tempDir, recursive: true);
            }
        }
    }

    [TestMethod]
    public void StartCommandRejectsUnsupportedCapacity()
    {
        using var backend = new FakeBackendController();
        using var viewModel = new LauncherViewModel(backend) { CapacityGb = 257.0 };

        Assert.IsFalse(viewModel.StartCommand.CanExecute(null));
        Assert.AreEqual("Choose a capacity from greater than 0 through 256 GB.", viewModel.ValidationMessage);
    }

    [TestMethod]
    public void BackendExitResetsViewModelState()
    {
        using var backend = new FakeBackendController();
        using var viewModel = new LauncherViewModel(backend);

        viewModel.Start();
        backend.RaiseExited();

        Assert.IsFalse(viewModel.IsRunning);
        Assert.AreEqual("Stopped", viewModel.Status);
    }

    private sealed class FakeBackendController : IBackendController
    {
        public event EventHandler<string>? LogReceived;
        public event EventHandler? Ready;
        public event EventHandler? Exited;

        public bool IsRunning { get; private set; }
        public int StartCalls { get; private set; }
        public int StopCalls { get; private set; }
        public BackendSettings? LastSettings { get; private set; }

        public void Start(BackendSettings settings)
        {
            IsRunning = true;
            StartCalls++;
            LastSettings = settings;
        }

        public void Stop()
        {
            IsRunning = false;
            StopCalls++;
        }

        public void RaiseReady()
        {
            Ready?.Invoke(this, EventArgs.Empty);
        }

        public void RaiseExited()
        {
            IsRunning = false;
            Exited?.Invoke(this, EventArgs.Empty);
        }

        public void Dispose()
        {
        }

        public void RaiseLog(string line)
        {
            LogReceived?.Invoke(this, line);
        }
    }
}

File: launcher/ClatterDrive.MacLauncher/Tests/ClatterDriveMacCoreTests/LauncherViewModelTests.swift
import XCTest
@testable import ClatterDriveMacCore

final class LauncherViewModelTests: XCTestCase {
    @MainActor
    func testStartUsesCurrentSettingsAndReadyCallback() async {
        let backend = FakeBackend()
        let model = LauncherViewModel(backend: backend)
        model.portText = "9099"
        model.backingDirectory = "/tmp/clatterdrive-test"
        model.capacityText = "32"
        model.filesystemProfile = "ext4_like"
        model.statePath = "/tmp/clatterdrive-test-state.json"

        model.start()
        backend.onReady?()
        await Task.yield()

        XCTAssertEqual(backend.startedSettings?.port, 9099)
        XCTAssertEqual(backend.startedSettings?.capacityGb, 32)
        XCTAssertEqual(backend.startedSettings?.filesystemProfile, "ext4_like")
        XCTAssertEqual(backend.startedSettings?.statePath, "/tmp/clatterdrive-test-state.json")
        XCTAssertTrue(model.isRunning)
        XCTAssertEqual(model.status, "Running")
    }

    @MainActor
    func testInvalidCapacityDoesNotStart() {
        let backend = FakeBackend()
        let model = LauncherViewModel(backend: backend)
        model.capacityText = "not-a-number"

        model.start()

        XCTAssertNil(backend.startedSettings)
        XCTAssertEqual(model.status, "Choose a capacity from greater than 0 through 256 GB.")
    }

    @MainActor
    func testInvalidPortDoesNotStart() {
        let backend = FakeBackend()
        let model = LauncherViewModel(backend: backend)
        model.portText = "abc"

        model.start()

        XCTAssertNil(backend.startedSettings)
        XCTAssertEqual(model.status, "Choose a port from 1 to 65535.")
    }
}

private final class FakeBackend: BackendControlling {
    var onLog: ((String) -> Void)?
    var onReady: (() -> Void)?
    var onExit: ((Int32?) -> Void)?
    var isRunning = false
    var startedSettings: BackendSettings?

    func start(settings: BackendSettings) throws {
        startedSettings = settings
        isRunning = true
    }

    func stop() {
        isRunning = false
    }

    func loadProfiles() throws -> ProfileCatalog {
        ProfileCatalog.fallback
    }
}
